configure.ac

```diff
@@ -42,16 +42,48 @@ if test "x$have_nanotime" = xyes; then
 fi
 AC_MSG_RESULT($have_nanotime)
 
+dnl Example of default-enabled feature
+AC_ARG_ENABLE([openmp],
+	AS_HELP_STRING([--disable-openmp], [Disable openMP capability]))
+useomp=no;
+AS_IF([test "x$enable_openmp" != xno], [
+  dnl Do the stuff needed for enabling the feature
+  AC_OPENMP
+  if test "$ac_cv_prog_c_openmp" = unsupported; then
+    AC_MSG_ERROR([OpenMP not supported])
+  else
+    useomp=yes;
+    AC_DEFINE([THREADED_OPENMP], [1], [openmp support is present])
 # Does the user want to turn on nested OMP? Enable only if you will be
 # profiling code regions inside of doubly-nested OpenMP regions. Using
 # --enable-nestedomp can increase the cost of getting the thread
 # number by about 50% even in regions which are not doubly-nested.
-AC_MSG_CHECKING([whether nested OMP is to be enabled])
-AC_ARG_ENABLE([nestedomp], [AS_HELP_STRING([--enable-nestedomp],
-              [build with nested OMP capability])])
-test "x$enable_nestedomp" = xyes || enable_nestedomp=no
+    AC_MSG_CHECKING([whether nested OMP is to be enabled])
+    AC_ARG_ENABLE([nestedomp], [AS_HELP_STRING([--enable-nestedomp],
+                  [build with nested OMP capability])])
+    test "x$enable_nestedomp" = xyes || enable_nestedomp=no
+    AC_MSG_RESULT($enable_nestedomp)
+  fi
+])
 AM_CONDITIONAL(ENABLE_NESTEDOMP, [test x$enable_nestedomp = xyes])
-AC_MSG_RESULT($enable_nestedomp)
+
+# Check for pthread library only if --disable-openmp was specified
+AS_IF([test "$useomp" = no], [
+dnl Example of default-disabled feature
+  AC_ARG_ENABLE([pthread],
+      AS_HELP_STRING([--enable-pthread], [Enable pthread capability]))
+
+  AS_IF([test "x$enable_pthread" = "xyes"], [
+    dnl Do the stuff needed for enabling the feature
+    AC_MSG_CHECKING([for pthread support])
+    AC_CHECK_LIB([pthread], [pthread_mutex_init])
+    if test "x$ac_cv_lib_pthread_pthread_mutex_init" = xyes; then
+       AC_DEFINE([THREADED_PTHREADS], [1], [pthreads library is present])
+    fi
+    AC_MSG_RESULT([$enable_pthread])
+  ])
+  
+])
```

Comment:
Things work a lot better if you check all enables/disables unconditionally. That is, don't have an option that only appears if another option has been specified, it's too confusing and hard to maintain as it grows.

So evaluate the --enable-pthread even if useomp was not specified.

Then, after all --enables/disables are processed, you can use some simple shell if statements to error out if incompatible options are enabled.

For each option, it's helpful if there is always a shell variable defined to either yes or no. For example:

```
# Does the user want to disable pnetcdf?
AC_MSG_CHECKING([whether pnetcdf is to be used])
AC_ARG_ENABLE([pnetcdf],
              [AS_HELP_STRING([--disable-pnetcdf],
                              [Disable pnetcdf use.])])
test "x$enable_pnetcdf" = xno || enable_pnetcdf=yes
AC_MSG_RESULT([$enable_pnetcdf])
AM_CONDITIONAL(BUILD_PNETCDF, [test "x$enable_pnetcdf" = xyes])

```
After this code is run, $enable_pnetcdf is set to "yes" or "no". So later in the script I can always do:

```
if test "x$enable_pnetcdf" = xyes; then
   # Do some stuff for pnetcdf builds.
fi

```

You do not need to use the AS_IF macro.